server/src/util/my-repos.ts

```diff
@@ -0,0 +1,150 @@
+/*
+ * Copyright 2017 Google Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+import gql from 'graphql-tag';
+import {GitHub} from '../github';
+import {MyReposQuery, MyReposQueryVariables} from '../gql-types';
+
+const MAX_STATS_RETRIES = 5;
+
+// Variables that impact the scoring algorithm.
+const SCORE_THRESHOLD = 20;
+const CONTRIBUTION_SIZE = 10;
+const CONTRIBUTION_WINDOW = 1000 * 60 * 60 * 24 * 90;  // 90 days.
+
+/**
+ * Returns a list of repos that the given user contributes to. It is ordered
+ * based on contribution recency and consistency.
+ */
+export async function getMyRepos(
+    github: GitHub, login: string, userToken: string): Promise<string[]> {
+  const repos = new Map<string, number>();
+  const results = github.cursorQuery<MyReposQuery, MyReposQueryVariables>(
+      reposQuery, {login}, (q) => q.user && q.user.contributedRepositories);
+  const promises = [];
+  for await (const data of results) {
+    if (data.user) {
+      for (const repo of data.user.contributedRepositories.nodes || []) {
+        if (repo) {
+          const promise = getContributionWeight(
+              github, repo.owner.login, repo.name, login, userToken);
+          promises.push(promise.then((score) => {
+            repos.set(repo.owner.login + '/' + repo.name, score);
+          }));
+        }
+      }
+    }
+  }
+  await Promise.all(promises);
+  const comparator = (a: string, b: string) => {
+    return (repos.get(a) || 0) < (repos.get(b) || 0) ? 1 : -1;
+  };
+  let result = Array.from(repos.keys());
+  // Filter out repos with small levels of contributions.
+  result = result.filter((key) => (repos.get(key) || 0) > SCORE_THRESHOLD);
+  return result.sort(comparator);
+}
+
+/**
+ * Given a repo and a user, returns a weighted contribution score for the repo
+ * (0-100).
+ */
+function getContributionWeight(
```

Comment:
Make this an `async` function? Looks like it should work fine that way, but maybe I missed something. To reject, just `throw`.

Reply:
Oh, sg. Done.